src/LinAlg/hiopLinSolverSparsePARDISO.cpp

```diff
@@ -0,0 +1,523 @@
+    int nrhs = 1;
+  
+    pardiso_d(pt_ , &maxfct_, &mnum_, &mtype_, &phase,
+              &n_, kVal_, kRowPtr_, jCol_,
+              NULL, &nrhs, 
+              iparm_, &msglvl_,
+              
+              drhs, dx, &error_, dparm_);
+
+    if ( error_ != 0) {
+      printf ("PardisoSolver - ERROR during backsolve: %d\n", error_ );
+      assert(false);
+    }
+
+    nlp_->runStats.linsolv.tmTriuSolves.stop();
+    return 1;
+  }
+
+
+  /*
+  *  PARDISO for unsymmetric sparse matrix
+  */
+  hiopLinSolverNonSymSparsePARDISO::hiopLinSolverNonSymSparsePARDISO(const int& n, const int& nnz, hiopNlpFormulation* nlp)
+    : hiopLinSolverNonSymSparse(n, nnz, nlp),
+    kRowPtr_{nullptr}, jCol_{nullptr}, kVal_{nullptr},
+    nFakeNegEigs_{-1},
+    rhs_{nullptr},
+    n_{n}, nnz_{-1}, is_initialized_{false}
+  {
+    maxfct_ = 1; //max number of fact having same sparsity pattern to keep at the same time
+    mnum_ = 1;   //actual matrix (as in index from 1 to maxfct)
+    msglvl_ = 0; //messaging level
+    mtype_ = 11; //real and unsymmetric
+    solver_ = 0; //sparse direct solver
+  }
+
+  hiopLinSolverNonSymSparsePARDISO::~hiopLinSolverNonSymSparsePARDISO()
+  {
+    /* Termination and release of memory */
+    int phase = -1; /* Release internal memory . */
+    pardiso_d(pt_ , &maxfct_, &mnum_, &mtype_, &phase,
+              &n_, kVal_, kRowPtr_, jCol_,
+              NULL, NULL, 
+              iparm_, &msglvl_, NULL, NULL, &error_, dparm_);
+
+    if(kRowPtr_)
+      delete [] kRowPtr_;
+    if(jCol_)
+      delete [] jCol_;
+    if(kVal_)
+      delete [] kVal_;
+    if(index_covert_CSR2Triplet_)
+      delete [] index_covert_CSR2Triplet_;
+    if(index_covert_extra_Diag2CSR_)
+      delete [] index_covert_extra_Diag2CSR_;
+    
+    if(rhs_) {
+      delete rhs_;
+    }
+    
+  }
+
+  void hiopLinSolverNonSymSparsePARDISO::firstCall()
+  {
+    assert(n_==M.n() && M.n()==M.m());
+    assert(n_>0);
+
+    // transfer triplet form to CSR form
+    // note that input is in lower triangular triplet form. First part is the sparse matrix, and the 2nd part are the additional diagonal elememts
+    // the 1st part is sorted by row
+
+    M.convertToCSR(nnz_, &kRowPtr_, &jCol_, &kVal_, &index_covert_CSR2Triplet_, &index_covert_extra_Diag2CSR_, extra_diag_nnz_map);
+
+    // need Fortran indexes
+    for( int i = 0; i < n_+1; i++) {
+      kRowPtr_[i] += 1;
+    }
+    for( int i = 0; i < nnz_; i++) {
+      jCol_[i] += 1;
+    }
+
+    /* initialize PARDISO */
+    pardisoinit_d(pt_, &mtype_, &solver_, iparm_, dparm_, &error_); 
+    if (error_!=0) {
+      std::cout << "PardisoSolver ERROR during pardisoinit:" << error_ << "." << std::endl;
+      assert(false);
+    }
+
+    /* Numbers of processors, value of OMP_NUM_THREADS */
+    char *var = getenv("OMP_NUM_THREADS");
+    if(var != NULL) {
+      sscanf( var, "%d", &num_threads_ );
+    } else {
+      printf("Set environment OMP_NUM_THREADS before using PARDISO.");
```

Comment:
wouldn't be better to just let the solver work with 1 thread (assuming it does not give an error if the `OMP_NUM_THREADS` is not set)?